cabal-install/Distribution/Client/IndexUtils.hs

```diff
@@ -561,40 +605,102 @@ packageListFromCache mkPkg hnd Cache{..} mode = accum mempty [] mempty cacheEntr
 -- Index cache data structure
 --
 
+-- | Read the 'Index' cache from the filesystem
+readIndexCache :: Index -> IO Cache
+readIndexCache index
+  | is01Index index = decodeFile (cacheFile index)
+  | otherwise       = liftM read00IndexCache $ BSS.readFile (cacheFile index)
+
+-- | Write the 'Index' cache to the filesystem
+writeIndexCache :: Index -> Cache -> IO ()
+writeIndexCache index cache
+  | is01Index index = encodeFile (cacheFile index) cache
+  | otherwise       = writeFile (cacheFile index) (show00IndexCache cache)
+
+-- | Cabal caches various information about the Hackage index
+data Cache = Cache {
+    cacheEntries :: [IndexCacheEntry]
+  }
+
+instance NFData Cache where
+    rnf = rnf . cacheEntries
+
 -- | Tar files are block structured with 512 byte blocks. Every header and file
 -- content starts on a block boundary.
 --
-type BlockNo = Tar.TarEntryOffset
-
-data IndexCacheEntry = CachePackageId PackageId BlockNo
-                     | CacheBuildTreeRef BuildTreeRefType BlockNo
-                     | CachePreference Dependency
-  deriving (Eq)
+type BlockNo = Word32 -- Tar.TarEntryOffset
```

Comment:
Hmm? Wouldn't it be better the other way around?

type BlockNo = Tar.TarEntryOffset -- Word32